initial commit of drupal doc for upsun (#4382)
platformsh/platformsh-docs: sites/upsun/src/get-started/stacks/drupal.md

---
title: Deploy Drupal on {{% vendor/name %}}
sidebarTitle: Drupal
sectionBefore: PHP
#layout: single
weight: -67
description: |
    Complete the last required steps to successfully deploy Drupal on {{% vendor/name %}}.
---

{{< note title="Note" theme="info" >}}
Before you start, check out the [{{% vendor/name %}} demo app](https://console.upsun.com/projects/create-project) and the main [Getting started guide](/get-started/here/_index.md).

> They provide all of the core concepts and common commands you need to know before using the materials below.

Reviewer: ```suggestion
They provide all of the core concepts and common commands you need to know before using the materials below.

It should also be noted that this guide works for the following variations of Drupal:

- ([v10.x](https://www.drupal.org/project/drupal/releases/10.3.12)

- [v11.x](https://www.drupal.org/project/drupal/releases/11.1.2)

- [CMS](https://new.drupal.org/docs/drupal-cms/get-started/install-drupal-cms))

```

{{< /note >}}

In general this guide should work for most variations of Drupal (
[v10.x](https://www.drupal.org/project/drupal/releases/10.3.12),

> [v11.x](https://www.drupal.org/project/drupal/releases/11.1.2),

Reviewer: ```suggestion

```

> [CMS](https://new.drupal.org/docs/drupal-cms/get-started/install-drupal-cms)). With that mind, it assumes you already

Reviewer: ```suggestion

```

> have the Drupal files in a repository, have a local copy of that repository, and selected both `Redis` and `MariaDB`

Reviewer: ```suggestion

```

> during the [Configure your project](/get-started/here/configure) portion of the

Reviewer: ```suggestion

```

> [Getting Started](/get-started/here) guide.

Reviewer: ```suggestion

```

For Drupal to successfully deploy and operate, **after completing the [Getting started guide](/get-started/here)**,
you still need to make a few changes to your {{% vendor/name %}} configuration.

{{% guides/requirements name="Drupal" %}}

## Configure

Open the `.upsun/config.yaml` file that's been generated and replace with the following:

```yaml {filename=".upsun/config.yaml"}
applications:
    drupal:
        type: "php:8.3"
        relationships:
            mariadb: 'db:mysql'
            redis: 'cache:redis'
        mounts:
            # The default Drupal files directory.

>             '/web/sites/default/files':

Author: I think the longer format, while it does increase the page length, is easier to read. @Kemi-Elizabeth do you have any insight/input?

>             '/web/sites/default/files':

Reviewer: When you say shortened version, do you mean without the comments? I found it quite easy to read but I _was_ taking my time. I understand that it might not be scannable, but do we want this to be easily scannable? I would think that people taking the time to read this article would want to take in all the details even if it makes the page a bit longer?

                            scripts: false
                        '^/sites/[^/]+/settings.*?\.php$':
                            scripts: false
                '/sites/default/files':
                    allow: true
                    expires: 5m
                    passthru: '/index.php'
                    root: 'web/sites/default/files'
                    scripts: false
                    rules:
                        '^/sites/default/files/(css|js)':
                            expires: 2w
        dependencies:
            php:
                composer/composer: "^2.7"
        hooks:
            build: |
                set -e
            # fast.
            deploy: |
                set -e
                php ./drush/upsun_generate_drush_yml.php
                cd web
                bash $PLATFORM_APP_DIR/drush/upsun_deploy_drupal.sh
        crons:
            # Run Drupal's cron tasks every 19 minutes.
            drupal:
                spec: '*/19 * * * *'
                commands:
                    start: 'cd web ; drush core-cron'
        runtime:
            # Enable the redis extension so Drupal can communicate with the Redis cache.

>             extensions:

Reviewer 2: ```
            extensions: { redis, sodium, apcu, blackfire }
```
?

                - redis
                - sodium
                - apcu
                - blackfire
        source:
            root: /
services:
    db:
        type: mariadb:10.6
    cache:
        type: redis:7.2
routes:
    "https://{default}/":
        type: upstream
        upstream: "drupal:http"
        cache:
            enabled: true
            # Base the cache on the session cookie and custom Drupal cookies. Ignore all other cookies.
            cookies: ['/^SS?ESS/', '/^Drupal.visitor/']
    "https://www.{default}/":
        type: redirect
        to: "https://{default}/"
```

> This configuration is identical to what we've recommended for deploying

Reviewer: ```suggestion

```

> [Drupal on Platform.sh](https://docs.platform.sh/guides/drupal/deploy.html), just

Reviewer: ```suggestion

```

> [updating for the few differences seen in Upsun's configuration](https://docs.upsun.com/learn/tutorials/migrating/from-psh.html).

Reviewer: ```suggestion

```

> 

Reviewer: ```suggestion
This configuration is similar to the deployment process for [Drupal on Platform.sh](https://docs.platform.sh/guides/drupal/deploy.html), however it is slightly updated for [Upsun's configuration](https://docs.upsun.com/learn/tutorials/migrating/from-psh.html). 
```

## Variables

> Next, the `project:init` command created a `.environment` file for us, containing environment variables for our two

Reviewer: @gilzow I think this might work a bit better? (updated the suggestion above ^)

services (MariaDB and Redis). Append the following Drush configuration to the bottom of that file:

```bash {location=".environment"}
# Allow executable app dependencies from Composer to be run from the path.
if [ -n "$PLATFORM_APP_DIR" -a -f "$PLATFORM_APP_DIR"/composer.json ] ; then
  bin=$(composer config bin-dir --working-dir="$PLATFORM_APP_DIR" --no-interaction 2>/dev/null)
  export PATH="${PLATFORM_APP_DIR}/${bin:-vendor/bin}:${PATH}"
fi
```

## `settings.php`
Open `web/sites/default/settings.php` and append the following to the bottom of that file.

```php {location="web/sites/default/settings.php"}
// Upsun configuration
if (getenv('PLATFORM_APPLICATION') && file_exists(__DIR__ . '/settings.upsun.php')) {
  include __DIR__ . '/settings.upsun.php';
}
```

## Upsun-specific settings
Then create a new Upsun-specific settings file `web/sites/default/settings.upsun.php` that leverages the variables

> defined in `.environment` that contains the following:

Reviewer: ```suggestion
defined in `.environment`. This file should contain the following:
```

  }
  if (!isset($settings['php_storage']['twig'])) {
    $settings['php_storage']['twig']['directory'] = $settings['file_private_path'];
  }

  // Set the project-specific entropy value, used for generating one-time
  // keys and such.
  $settings['hash_salt'] = empty($settings['hash_salt']) ? getenv('PLATFORM_PROJECT_ENTROPY') : $settings['hash_salt'];

  // Set the deployment identifier, which is used by some Drupal cache systems.
  $settings['deployment_identifier'] = $settings['deployment_identifier'] ?? getenv('PLATFORM_TREE_ID');;
}

// The 'trusted_hosts_pattern' setting allows an admin to restrict the Host header values
// that are considered trusted.  If an attacker sends a request with a custom-crafted Host
// header then it can be an injection vector, depending on how the Host header is used.
// However, Platform.sh already replaces the Host header with the route that was used to reach
// Platform.sh, so it is guaranteed to be safe.  The following line explicitly allows all
// Host headers, as the only possible Host header is already guaranteed safe.
$settings['trusted_host_patterns'] = ['.*'];
```

## `config/sync`

> Create the `config/sync` empty directory referenced in this settings file:

Reviewer: ```suggestion
Create the `config/sync` empty directory referenced in the following settings file:
```

Author: ```suggestion
Create the `config/sync` empty directory referenced in the settings file:
```

```bash {location="Terminal"}
mkdir -p config/sync && touch config/sync/.gitkeep
```

## Configuration reader

> Install the helpful [Config Reader library](https://github.com/platformsh/config-reader-php), which is required, and

Reviewer: ```suggestion
Install the required [Config Reader library](https://github.com/platformsh/config-reader-php).
```

> will help us to pull routing details for each environment into our settings (highlighted in the snippet in the next step).

Reviewer: ```suggestion
This will help us to pull routing details for each environment into our settings (highlighted in the snippet in the next step).
```

```bash
composer require platformsh/config-reader
```
## Drush

> Lastly, we need to create some files referenced in `.upsun/config.yaml` to configure Drush so it can be used within the

Reviewer: ```suggestion
To configure Drush, use the following command to create the files that will be referenced in the configuration process. This process will allow Drush to be used within the Upsun container.
```

> Upsun container.

Reviewer: ```suggestion

```

```bash {location="Terminal"}
mkdir drush && touch drush/upsun_deploy_drupal.sh && touch drush/upsun_generate_drush_yml.php
```

Fill out the `drush/upsun_deploy_drupal.sh` with the following:

```bash {location="drush/upsun_deploy_drupal.sh"}
#!/usr/bin/env bash
#
# We don't want to run drush commands if drupal isn't installed.
# Similarly, we don't want to attempt to run config-import if there aren't any config files to import

> 

Reviewer: I've tried to remove all uses of first person plural and tried to tone down the conversational TOV so that it fits more with the formality of the Docs as opposed to the DevCenter articles. I did see that some of the comments within the code are a bit colloquial (ie. the one above uses a tone that could come across slightly conversational). Do we want to keep those as they are (I'm not sure if changing the comments might affect the code snippet example)? @gilzow

Author: @Kemi-Elizabeth  i think the comments are important so the reader know what the commands are doing. We could change it to be:
```
# Don't run drush commands if drupal isn't installed.
# Don't run config-import if there aren't any config files to import
```
?

Reviewer: Perfect! Yes, I still think the comments are necessary, I just wasn't sure if we could change the wording. But that's perfect and definitely along the lines of what I was thinking.

                commands:
                    start: 'cd web ; drush core-cron'
        runtime:
            # Enable the redis extension so Drupal can communicate with the Redis cache.
            extensions:
                - redis
                - sodium
                - apcu
                - blackfire
        source:
            root: /
services:
    db:
        type: mariadb:10.6
    cache:
        type: redis:7.2
routes:
    "https://{default}/":
        type: upstream
        upstream: "drupal:http"
        cache:
            enabled: true
            # Base the cache on the session cookie and custom Drupal cookies. Ignore all other cookies.
            cookies: ['/^SS?ESS/', '/^Drupal.visitor/']
    "https://www.{default}/":
        type: redirect
        to: "https://{default}/"
```

This configuration is similar to the deployment process for [Drupal on Platform.sh](https://docs.platform.sh/guides/drupal/deploy.html), however it is slightly updated for [Upsun's configuration](https://docs.upsun.com/learn/tutorials/migrating/from-psh.html). 
## Variables

Next, the `project:init` command created a `.environment` file for us, containing environment variables for our two

> services (MariaDB and Redis). Append the following Drush configuration to the bottom of that file:

Reviewer: ```suggestion

```

// Platform.sh, so it is guaranteed to be safe.  The following line explicitly allows all
// Host headers, as the only possible Host header is already guaranteed safe.
$settings['trusted_host_patterns'] = ['.*'];
```

## `config/sync`
Create the `config/sync` empty directory referenced in the settings file:

```bash {location="Terminal"}
mkdir -p config/sync && touch config/sync/.gitkeep
```

## Configuration reader

Install the required [Config Reader library](https://github.com/platformsh/config-reader-php).
This will help us to pull routing details for each environment into our settings (highlighted in the snippet in the next step).

```bash
composer require platformsh/config-reader
```
## Drush
To configure Drush, use the following command to create the files that will be referenced in the configuration process. This process will allow Drush to be used within the Upsun container.

```bash {location="Terminal"}
mkdir drush && touch drush/upsun_deploy_drupal.sh && touch drush/upsun_generate_drush_yml.php
```

Fill out the `drush/upsun_deploy_drupal.sh` with the following:

```bash {location="drush/upsun_deploy_drupal.sh"}
#!/usr/bin/env bash
#

> # We don't want to run drush commands if drupal isn't installed.

Author: ```suggestion
# Don't run drush commands if drupal isn't installed.
```

> # Similarly, we don't want to attempt to run config-import if there aren't any config files to import

Author: ```suggestion
# Don't run config-import if there aren't any config files to import
```
